File: inspector/mos.cpp
#include "mos.h"
#include "mainwindow.h"
#include <QLabel>
#include <QDebug>
#include <zlib.h>

MOS::MOS(QSharedPointer<Handle> handle, QWidget *parent) : QScrollArea(parent) {
  QString key = handle->r4();
  if (key == "MOSC") {
    ulong len = handle->r32();
    handle->skip(4); //unknown
    Bytef *in = (Bytef *)handle->read(handle->length() - 12);
    Bytef *out = new Bytef[len];
    uncompress(out, &len, in, handle->length() - 12);
    QByteArray arr((const char *)out, len);
    handle = QSharedPointer<Handle>(new Handle(arr));
    key = handle->r4();
  }

  if (key != "MOS ")
    throw FileError("Not a MOS file");

  handle->skip(4); //unknown
  quint16 w = handle->r16();
  quint16 h = handle->r16();
  quint16 cols = handle->r16();
  quint16 rows = handle->r16();
  handle->skip(4); //bsize
  quint32 offset = handle->r32();

  QImage image(w, h, QImage::Format_RGBA8888);
  uchar *bits = image.bits();
  int stride = image.bytesPerLine();

  handle->seek(offset);
  quint8 *palette = (quint8 *)handle->read(256 * 4 * cols * rows);
  quint32 *offsets = new quint32[cols * rows];
  for (int y = 0, off = 0; y < rows; y++)
    for (int x = 0; x < cols; x++, off++)
      offsets[off] = handle->r32();
  qint64 tiledata = handle->tell();
  int poff = 0;
  quint8 c;
  for (int ty = 0, off = 0; ty < rows; ty++) {
    for (int tx = 0; tx < cols; tx++, off++) {
      quint16 tw = 64, th = 64;
      if (tx == cols - 1) tw = w & 63;
      if (ty == rows - 1) th = h & 63;
      if (tw == 0) tw = 64;
      if (th == 0) th = 64;
      handle->seek(tiledata + offsets[off]);

      for (int y = 0; y < th; y++) {
        int out = (y + ty * 64) * stride + tx * 64 * 4;
        for (int x = 0; x < tw; x++) {
          c = handle->r8();
          bits[out++] = palette[poff + c * 4 + 2];
          bits[out++] = palette[poff + c * 4 + 1];
          bits[out++] = palette[poff + c * 4 + 0];
          if (palette[poff + c * 4 + 2] == 0 &&
              palette[poff + c * 4 + 1] == 0xff &&
              palette[poff + c * 4 + 0] == 0)
            bits[out++] = 0;
          else
            bits[out++] = 255;
        }
      }
      poff += 256 * 4;
    }
  }
  delete[] offsets;

  QLabel *label = new QLabel(this);
  label->setPixmap(QPixmap::fromImage(image));
  setWidget(label);
}

File: greenland/mos.go
/** @copyright 2025 Sean Kasun */
package main

import (
	"bytes"
	"compress/zlib"
	"image"
	"image/png"
	"io"

	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/canvas"
	"fyne.io/fyne/v2/container"
)

type MOS struct {
	view  fyne.CanvasObject
	image image.Image
	file  *File
}

func NewMOS(res *File) ResView {
	key := res.R4()
	if key == "MOSC" {
		res.Skip(4) // "V1  "
		length := res.R32()
		in := bytes.NewBuffer(res.Read(res.Length() - res.Tell()))
		reader, err := zlib.NewReader(in)
		if err != nil {
			return NewErr(err.Error())
		}
		arr, err := io.ReadAll(reader)
		if err != nil {
			return NewErr(err.Error())
		}
		if len(arr) != length {
			return NewErr("Invalid size")
		}
		res = NewFile(arr)
		key = res.R4()
	}

	if key != "MOS " {
		return NewErr("Not a MOS file")
	}

	res.Skip(4) // "V1  "
	w := res.R16()
	h := res.R16()
	cols := res.R16()
	rows := res.R16()
	res.Skip(4) // bsize
	offset := res.R32()

	img := image.NewRGBA(image.Rect(0, 0, w, h))

	res.Seek(offset)
	palette := res.Read(256 * 4 * cols * rows)
	offsets := make([]int, cols*rows)
	for i := range cols * rows {
		offsets[i] = res.R32()
	}
	tiledata := res.Tell()
	off := 0
	for ty := range rows {
		th := 64
		if ty == rows-1 {
			th = h & 63
		}
		if th == 0 {
			th = 64
		}
		for tx := range cols {
			tw := 64
			if tx == cols-1 {
				tw = w & 63
			}
			if tw == 0 {
				tw = 64
			}
			res.Seek(tiledata + offsets[off])

			for y := range th {
				out := (y+ty*64)*img.Stride + tx*64*4
				for range tw {
					c := res.R8()
					r := palette[off*256*4+c*4+2]
					g := palette[off*256*4+c*4+1]
					b := palette[off*256*4+c*4]
					img.Pix[out] = r
					img.Pix[out+1] = g
					img.Pix[out+2] = b
					if r == 0 && g == 0xff && b == 0 {
						img.Pix[out+3] = 0
					} else {
						img.Pix[out+3] = 0xff
					}
					out += 4
				}
			}
			off++
		}
	}

	i := canvas.NewImageFromImage(img)
	i.FillMode = canvas.ImageFillOriginal
	i.ScaleMode = canvas.ImageScalePixels
	return &MOS{container.NewCenter(i), img, res}
}

func (m *MOS) Gestalt() Gestalt {
	return CanSave | CanExport
}

func (m *MOS) View() fyne.CanvasObject {
	return m.view
}

func (m *MOS) Save(w io.Writer) {
	w.Write(m.file.data)
}

func (m *MOS) Export(w io.Writer) {
	png.Encode(w, m.image)
}
